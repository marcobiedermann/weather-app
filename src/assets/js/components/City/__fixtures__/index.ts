const city = {
  dt: 0,
  id: 1,
  main: {
    temp: 0,
  },
  name: '',
  weather: [
    {
      description: '',
      icon: '',
      id: 1,
    },
  ],
};

export default city;
