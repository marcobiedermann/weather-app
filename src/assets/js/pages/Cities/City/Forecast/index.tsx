import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import Loader from '../../../../components/Loader';

interface Params {
  cityId: string;
}

interface ForecastPageQuery {
  city: City;
  cnt: number;
  cod: string;
  list: List[];
  message: number;
}

interface City {
  coord: Coord;
  country: string;
  id: number;
  name: string;
  population: number;
  timezone: number;
}

interface Coord {
  lat: number;
  lon: number;
}

interface List {
  clouds: number;
  deg: number;
  dt: number;
  feels_like: FeelsLike;
  humidity: number;
  pop: number;
  pressure: number;
  speed: number;
  sunrise: number;
  sunset: number;
  temp: Temp;
  weather: Weather[];
}

interface FeelsLike {
  day: number;
  eve: number;
  morn: number;
  night: number;
}

interface Temp {
  day: number;
  eve: number;
  max: number;
  min: number;
  morn: number;
  night: number;
}

interface Weather {
  description: string;
  icon: string;
  id: number;
  main: string;
}

const ForecastPage: FC = () => {
  const { cityId } = useParams<Params>();
  const { data, error } = useSWR<ForecastPageQuery>(`/forecast/daily?id=${cityId}`);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Forecast</title>
      </Helmet>
      <ul>
        {data.list.map((item) => (
          <li key={item.dt}>
            <div>Min: {item.temp.min}</div>
            <div>Max: {item.temp.max}</div>
            <div>
              {item.weather[0].description}
              <img
                src={`https://openweathermap.org/img/w/${item.weather[0].icon}.png`}
                alt={item.weather[0].description}
              />
            </div>
          </li>
        ))}
      </ul>
    </>
  );
};

export default ForecastPage;
