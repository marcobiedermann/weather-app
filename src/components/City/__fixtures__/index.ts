const city = {
  clouds: {
    all: 0,
  },
  coord: {
    lon: 13.41,
    lat: 52.52,
  },
  dt: 1605434959,
  id: 2950159,
  main: {
    feels_like: 9.39,
    humidity: 76,
    pressure: 1014,
    temp_max: 13.89,
    temp_min: 11,
    temp: 12.61,
  },
  name: 'Berlin',
  sys: {
    country: 'DE',
    sunrise: 1605421729,
    sunset: 1605453206,
    timezone: 3600,
  },
  weather: [
    {
      description: 'clear sky',
      icon: '01d',
      id: 800,
      main: 'Clear',
    },
  ],
  visibility: 10000,
  wind: {
    deg: 170,
    speed: 4.1,
  },
};

export default city;
