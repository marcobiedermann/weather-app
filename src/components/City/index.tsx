import format from 'date-fns/format';

interface Weather {
  description: string;
  icon: string;
  id: number;
}

interface CityProps {
  dt: number;
  main: {
    temp: number;
  };
  name: string;
  weather: Weather[];
}

function City(props: CityProps): JSX.Element {
  const { dt, main, name, weather, ...otherProps } = props;

  return (
    <div {...otherProps}>
      <h1>{name}</h1>
      {format(dt * 1000, 'yyyy-MM-dd hh:mm')}
      <div>Temp: {main.temp}</div>
      {weather && (
        <ul>
          {weather.map((entry) => (
            <li key={entry.id}>
              {entry.description}
              <img
                src={`https://openweathermap.org/img/w/${entry.icon}.png`}
                alt={entry.description}
              />
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}

export type { CityProps };
export default City;
