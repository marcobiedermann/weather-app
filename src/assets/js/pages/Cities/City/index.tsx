import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import City from '../../../components/City';
import Loader from '../../../components/Loader';

interface Params {
  cityId: string;
}

interface CityPageQuery {
  coord: Coord;
  weather: Weather[];
  base: string;
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  sys: Sys;
  timezone: number;
  id: number;
  name: string;
  cod: number;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  type: number;
  id: number;
  country: string;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

const CityPage: FC = () => {
  const { cityId } = useParams<Params>();
  const { data, error } = useSWR<CityPageQuery>(`/weather?id=${cityId}`);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>{data.name}</title>
      </Helmet>
      <City {...data} />
    </>
  );
};

export default CityPage;
