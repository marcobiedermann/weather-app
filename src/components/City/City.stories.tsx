import type { Meta, StoryObj } from '@storybook/react';
import City from '.';

const meta: Meta<typeof City> = {
  component: City,
  args: {
    weather: [
      {
        id: 800,
        description: 'clear sky',
        icon: '01d',
      },
    ],
    main: {
      temp: 25.98,
      temp_min: 22.77,
      temp_max: 28.88,
    },
    name: 'Berlin',
  },
};

type Story = StoryObj<typeof City>;

const Default: Story = {};

export { Default };
export default meta;
