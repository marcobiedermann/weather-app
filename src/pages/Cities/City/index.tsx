import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import City from '../../../components/City';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';

interface CityPageQuery {
  coord: Coord;
  weather: Weather[];
  base: string;
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  sys: Sys;
  timezone: number;
  id: number;
  name: string;
  cod: number;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  type: number;
  id: number;
  country: string;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

function CityPage(): JSX.Element {
  const { cityId } = useParams();
  const { data, error } = useSWR<CityPageQuery>(`/weather?id=${cityId}`);

  if (error) {
    return <Error message={error.message} />;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>{data.name}</title>
      </Helmet>
      <City {...data} />
    </>
  );
}

export default CityPage;
