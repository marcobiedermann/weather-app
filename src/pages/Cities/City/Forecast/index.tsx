import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import Error from '../../../../components/Error';
import Forecast from '../../../../components/Forecast';
import Loader from '../../../../components/Loader';
import { useDailyForeCast } from '../../../../hooks';

function ForecastPage(): JSX.Element {
  const { cityId } = useParams();
  const { data, error, isError, isLoading } = useDailyForeCast(parseInt(cityId!, 10));

  if (isError) {
    return <Error message={error.message} />;
  }

  if (isLoading) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Daily Forecast</title>
      </Helmet>
      <Forecast list={data?.list} />
    </>
  );
}

export default ForecastPage;
