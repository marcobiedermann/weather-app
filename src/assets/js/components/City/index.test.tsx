import React from 'react';
import renderer from 'react-test-renderer';
import City from '.';
import cityFixtures from './__fixtures__';

describe('components/City', () => {
  it('renders correctly', () => {
    expect.assertions(1);

    const tree = renderer.create(<City {...cityFixtures} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
