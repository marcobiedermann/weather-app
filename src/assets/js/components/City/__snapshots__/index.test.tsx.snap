// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/City renders correctly 1`] = `
<div
  id={1}
>
  <h1>
    
  </h1>
  1970-01-01 01:00
  <div>
    Temp: 
    0
  </div>
  <ul>
    <li>
      
      <img
        alt=""
        src="https://openweathermap.org/img/w/.png"
      />
    </li>
  </ul>
</div>
`;
